main/autogen/autoloader.cc

```diff
@@ -516,6 +528,19 @@ void AutoloadWriter::writeAutoloads(const core::GlobalState &gs, WorkerPool &wor
         }
         for (const auto &file : existingFilesSet) {
             FileOps::removeFile(file);
+
+            // Remove all empty directories along path. This prevents zeitwerk from vivifying
+            // dangling modules.
+            std::string_view filePath = file;
+            int curDirPos = filePath.find_last_of('/');
+            while (curDirPos > 0) {
+                const auto curDir = filePath.substr(0, curDirPos);
+                if (curDir == path || !FileOps::removeEmptyDir(curDir)) {
+                    break;
+                }
+
+                curDirPos = filePath.find_last_of('/', curDirPos - 1);
+            }
```

Comment:
Open to suggestions on writing this logic differently. In this new world, empty dirs have a minor runtime side-effect of generating dangling autoloads.

This would not be a problem in the latest version of Zeitwerk, which eagerly ignores empty dirs. However, it has its own perf unknowns which we haven't de-risked, so we can't bump to it just yet.